### ISSUE TITLE
Unavailable guilds dispatched through the wrong event

### ISSUE THREAD

**Issue by DeveloperAnonymous** [nextcord/nextcord#173]
### Summary

Unavailable guilds dispatched through guild_remove

### Reproduction Steps

<#889252941704343552> on discord

### Minimal Reproducible Code

```python
async def on_guild_remove(self, guild):
        await util.duelogger.info("BattleBanana has been removed from the guild **%s** (%s members)"
                                  % (util.ultra_escape_string(guild.name), guild.member_count))
```

### Expected Results

BattleBanana has been removed from the guild **BattleBanana** (1400 members)

### Actual Results

BattleBanana has been removed from the guild **None** (None members)

### Intents

default, members & guilds

### System Information

- Python v3.8.10-final
- nextcord v2.0.0-alpha
    - nextcord pkg_resources: v2.0.0a2
- aiohttp v3.6.3
- system info: Linux 5.4.0-84-generic #94-Ubuntu SMP Thu Aug 26 20:27:37 UTC 2021

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

<#889252941704343552> on discord

**Comment by Suven-p** [comment]
Has this been fixed? I couldn't replicate this.

**Comment by ooliver1** [comment]
i think i had this issue before too

**Comment by EmreTech** [comment]
Isn't this intentional? According to the [Discord API Docs](https://discord.com/developers/docs/topics/gateway#guild-delete), if a guild becomes unavailable, then the `GUILD_DELETE` event will be dispatched. Maybe what you want is for the unavailable guild event to be separated from the guild delete event?

**Comment by ooliver1** [comment]
That's very true; changing purpose of this issue:

**Comment by DeveloperAnonymous** [comment]
I see you've kinda overlooked @EmreTech's suggestion: `Maybe what you want is for the unavailable guild event to be separated from the guild delete event?`

Personally, I 100% agree with that, I don't think an unavailable guild should be dispatched as a deleted guild. Despite being easy to check and ignore, it'd be best to separate that because if you don't know it, it could very easily turn very badly.

An example with that would be my bot and I am pretty sure I'm probably not the only one doing that: Whenever a server is removed I purge all of its data from the database as they are no longer useful.

Of course this can be easily avoided, but at the first look of the event name `on_guild_remove`, you'll never guess it also dispatch unavailable guilds and you'll never really know until someone reports it\*, which is very unlikely, or reading the doc, which you'll most likely won't think its necessary because of the self-documenting name

\* Also quick side note on the report, users are normally pretty vague on their report, if any, such as "MY SERVER LOST ALL ITS THING" and then you'll be here wondering "Da heck? It's probably on them" because it never happened before. Other case would just be the user actually removing your bot because it lost everything and they're too upset to redo everything. 

Obviously, if you just have a bot where you only store the command key it doesn't matter, but in my case I hold a lot more information about a server, such as: Quests, Weapons, Muted channels (Where the bot won't send notifications), Blocked channels (Where the bot won't reply to commands) and Command Key. They can have up to 200 weapons and 200 quests. Imagine recreating that.

So you might think that "Oh, that's a you problem". I agree. So let's take another, more generic, example. Let's do it with a moderation bot, right? So what could we have in a moderation bot per-server? List of blacklisted words, List of 'administrator' roles, list of whitelisted roles to bypass the blacklisted words, list of whitelisted channels, maybe a custom warning message, custom kick message, custom ban message, custom mute message and so on. Just look at all the possible options Dyno offer, you need a whole web dashboard to configure your server since there's just so many options.

So for the sake of being transparent and avoid unnecessary/unexpected dispatches and/or confusion, I can only suggest to look into @EmreTech's suggestion of moving the unavailable guild event to its own

**Comment by ooliver1** [comment]
i too was confused when i got this until i remembered this issue

**Comment by EmreTech** [comment]
@ooliver1 don't forget to revert the title & description

**Comment by DenverCoder1** [comment]
I don't think it's the responsibility of the lib to differentiate events that Discord itself does not differentiate.

While I don't disagree that it could be useful, the `Guild Delete` event should work as the Discord Documentation describes it should.

If the user wants to differentiate them, they can do that themselves and there should be documentation telling them how to do that (similar to Discord's documentation).

**Comment by EmreTech** [comment]
Maybe the `GUILD_DELETE` event should have an unavailable parameter, but then again I think the Guild object should already have that attribute

**Comment by EmreTech** [comment]
I was about to mention this
So somehow the `guild_remove` event is still dispatched probably along with the `guild_unavailable` event even though the code explicitly tries to return after the `guild_unavailable` event dispatching.

**Comment by DenverCoder1** [comment]
It seems from the code that it works the way you expect it should. Do you have any logs or similar showing that there is a bug?

**Comment by EmreTech** [comment]
@DeveloperAnonymous I know it's a little difficult to reproduce this issue (since unavailable guilds are rare) but try to log the unavailable attribute of the guild parameter in your `on_guild_remove` event callback.

**Comment by ooliver1** [comment]
i can chip in on this, when i had an unavailable guild, it was dispatched to guild_remove, checking and shielding against 
```py
if guild.unavailable:
    return
``` stopped the constant logging of leaving a guild on every restart

**Comment by DenverCoder1** [comment]
This issue has been brought up many times in the dpy server as well. It seems like sometimes Discord sends the Guild Delete event without the unavailable field? In which case that's on Discord, not sure how that should be handled.

See thread "Unavailable guild in on_guild_remove on startup" [<#886973276541304832>](https://discord.com/channels/336642139381301249/886973276541304832)

**Comment by EmreTech** [comment]
Couldn't you just fetch the guild to ensure that it's available or not?

**Comment by ooliver1** [comment]
okay so i suppose ill revert this back to document it then

**Comment by ooliver1** [comment]
> Couldn't you just fetch the guild to ensure that it's available or not?

yes but discord dispatches it, we are a dapi wrapper after all, no need to make magic when discord does it themselves

**Comment by DeveloperAnonymous** [comment]
Well from what it says on the channel you linked, it seems that it's normally properly done, but some times, on discord ends, it doesn't set the unavailable attribute which would seem to be the root cause of this issue @DenverCoder1 @ooliver1

**Comment by DeveloperAnonymous** [comment]
And for your comment @EmreTech, it could take days or weeks to happen again and on top of that I switched to another library during the meantime
